File: paster-webapp/src/main/scala/org/apache/tiles/el/ScopeELResolver.scala
package org.apache.tiles.el

import jakarta.el.ELResolver
import org.apache.tiles.request.Request
import scala.jdk.CollectionConverters._
import jakarta.el.ELContext
import java.beans.FeatureDescriptor
import java.util
import java.util.Collections


/**
 * Resolves beans in request, session and application scope.
 *
 * @version $Rev$ $Date$
 * @since 2.2.1
 */
class ScopeELResolver extends ELResolver{


  /**
   * The length of the suffix: "Scope".
   */
  private val SUFFIX_LENGTH = 5

  /** {@inheritDoc } */
  def getCommonPropertyType(context: ELContext, base: Any): Class[_] = { // only resolve at the root of the context
    if (base != null)
        return null
    classOf[java.util.Map[_,_]]
  }

  def getFeatureDescriptors(context: ELContext, base: Any): java.util.Iterator[FeatureDescriptor] = {
    if (base != null) {
      val retValue:java.util.List[FeatureDescriptor] = Collections.emptyList
      return retValue.iterator
    }
    val list = new util.ArrayList[FeatureDescriptor]
    val request = context.getContext(classOf[Request]).asInstanceOf[Request]
    for (scope:String <- request.getAvailableScopes.asScala) {
      val descriptor = new FeatureDescriptor
      descriptor.setDisplayName(scope + "Scope")
      descriptor.setExpert(false)
      descriptor.setHidden(false)
      descriptor.setName(scope + "Scope")
      descriptor.setPreferred(true)
      descriptor.setShortDescription("")
      descriptor.setValue("type", classOf[java.util.Map[_,_]])
      descriptor.setValue("resolvableAtDesignTime", java.lang.Boolean.FALSE)
      list.add(descriptor)
    }
    list.iterator
  }

  def getType(context: ELContext, base: Any, property: Any): Class[_] = {
    if (base != null || property == null || !property.isInstanceOf[String]
      || !(property.asInstanceOf[String]).endsWith("Scope")) return null
    classOf[java.util.Map[_,_]]
  }

  def getValue(context: ELContext, base: Any, property: Any): Any = {
    if (base != null) return null
    var retValue:Object = null
    val propertyString = property.asInstanceOf[String]
    if (property.isInstanceOf[String] && propertyString.endsWith("Scope")) {
      val request = context.getContext(classOf[Request]).asInstanceOf[Request]
      retValue = request.getContext(propertyString
        .substring(0, propertyString.length - SUFFIX_LENGTH))
    }
    if (retValue != null) context.setPropertyResolved(true)
    retValue
  }

  def isReadOnly(context: ELContext, base: Any, property: Any): Boolean = {
    if (context == null) throw new NullPointerException
    true
  }

  def setValue(context: ELContext, base: Any, property: Any, value: Any): Unit = {
    // Does nothing for the moment.
  }
}

File: libs/tiles-jakarta-stripped/src/main/java/org/apache/tiles/el/ScopeELResolver.java
/*
 * $Id: ScopeELResolver.java 1049676 2010-12-15 19:38:54Z apetrelli $
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tiles.el;
import java.util.Map;

import jakarta.el.ELContext;
import jakarta.el.ELResolver;
import org.apache.tiles.request.Request;
/**
 * Resolves beans in request, session and application scope.
 *
 * @version $Rev: 1049676 $ $Date: 2010-12-16 06:38:54 +1100 (Thu, 16 Dec 2010) $
 * @since 2.2.1
 */
public class ScopeELResolver extends ELResolver {
    /**
     * The length of the suffix: "Scope".
     */
    private static final int SUFFIX_LENGTH = 5;
    /**
     * {@inheritDoc}
     */
    @Override
    public Class<?> getCommonPropertyType(ELContext context, Object base) {
        // only resolve at the root of the context
        return base != null ? null : Map.class;
    }
    /**
     * {@inheritDoc}
     */
    @Override
    public Class<?> getType(ELContext context, Object base, Object property) {
        return base != null || !(property instanceof String s)
                || !s.endsWith("Scope") ? null : Map.class;
    }
    /**
     * {@inheritDoc}
     */
    @Override
    public Object getValue(ELContext context, Object base, Object property) {
        if (base != null)
            return null;

        Object retValue = null;
        String propertyString = (String) property;
        if (property != null && propertyString.endsWith("Scope")) {
            Request request = (Request) context
                    .getContext(Request.class);
            retValue = request.getContext(propertyString.substring(0,
                    propertyString.length() - SUFFIX_LENGTH));
        }
        if (retValue != null)
            context.setPropertyResolved(true);

        return retValue;
    }
    /**
     * {@inheritDoc}
     */
    @Override
    public boolean isReadOnly(ELContext context, Object base, Object property) {
        if (context == null)
            throw new NullPointerException();

        return true;
    }
    /**
     * {@inheritDoc}
     */
    @Override
    public void setValue(ELContext context, Object base, Object property,
                         Object value) {
        // Does nothing for the moment.
    }
}
